src/main/java/org/takes/rq/RqRequestLine.java

```diff
@@ -0,0 +1,212 @@
+ * SOFTWARE.
+ */
+package org.takes.rq;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import lombok.EqualsAndHashCode;
+import org.takes.HttpException;
+import org.takes.Request;
+
+/**
+ * HTTP Request-Line parsing.
+ *
+ * <p>All implementations of this interface must be immutable and thread-safe.
+ *
+ * @author Vladimir Maksimenko (xupypr@xupypr.com)
+ * @version $Id$
+ * @since 0.29.1
+ */
+@SuppressWarnings("PMD.TooManyMethods")
+public interface RqRequestLine extends Request {
+
+    /**
+     * Get Request-Line header.
+     * @return HTTP Request-Line header
+     * @throws IOException If fails
+     */
+    String header() throws IOException;
+
+    /**
+     * Get Request-Line method token.
+     * @return HTTP Request-Line method token
+     * @throws IOException If fails
+     */
+    String method() throws IOException;
+
+    /**
+     * Get Request-Line Request-URI token.
+     * @return HTTP Request-Line method token
+     * @throws IOException If fails
+     */
+    String uri() throws IOException;
+
+    /**
+     * Get Request-Line HTTP-Version token.
+     * @return HTTP Request-Line method token
+     * @throws IOException If fails
+     */
+    String version() throws IOException;
+
+    /**
+     * Request decorator for Request-Line header validation
+     *
+     * <p>The class is immutable and thread-safe.
+     * @author Vladimir Maksimenko (xupypr@xupypr.com)
+     * @version $Id$
+     * @since 1.0
+     */
+    @EqualsAndHashCode(callSuper = true)
+    final class Base extends RqWrap implements RqRequestLine {
+        /**
+         * HTTP Request-line pattern.
+         * [!-~] is for method or extension-method token (octets 33 - 126).
+         * @see <a href="http://www.w3.org/Protocols/rfc2616/rfc2616-sec5.html#sec5.1">RFC 2616</a>
+         */
+        private static final Pattern PATTERN = Pattern.compile(
+            "([!-~]+) ([^ ]+)( [^ ]+)?"
+        );
+
+        private static enum Token {
+            /**
+             * METHOD token.
+             */
+            METHOD(1),
+
+            /**
+             * URI token.
+             */
+            URI(2),
+
+            /**
+             * HTTPVERSION token.
+             */
+            HTTPVERSION(3);
+
+            /**
+             * Value.
+             */
+            private final int value;
+
+            /**
+             * Ctor.
+             * @param val Value
+             */
+            private Token(final int val) {
+                this.value = val;
+            }
+        }
+
+        /**
+         * Ctor.
+         * @param req Original request
+         */
+        public Base(final Request req) {
+            super(req);
+        }
+
+        @Override
+        public String header() throws IOException {
+            final String line = this.line();
+            this.matcher(line);
+            return line;
```

Comment:
@xupyprmv maybe I'm wrong with this exact example, but `this.matcher(line)` is a code smell. You should **never** use functions as procedures, ignoring their returns.